### PR TITLE
#351: Add IFileSystem property to base classes

### PR DESCRIPTION
**Pull request by herebebeasties**
This allows people to get at the underlying IFileSystem implementation when developing extension methods.

For example, it may make sense to put your extension method on DirectoryBase, but you may require FileBase to actually implement it. This makes that possible.

In addition it fixes a bug in MockFileInfo.Replace where the returned FileInfo was for the concrete FileSystem not the mock one (bug caused by implicit casting).

I think this commit actually removes the need for supporting implicit casting at all.

### CODE REVIEW

**Comment by herebebeasties** [comment]
This implements #351.

File: System.IO.Abstractions.TestingHelpers/IMockFileDataAccessor.cs [hunk 1]
```diff
@@ -62,5 +62,7 @@ public interface IMockFileDataAccessor
         IDriveInfoFactory DriveInfo { get; }
 
         PathVerifier PathVerifier { get; }
+
+        IFileSystem FileSystem { get; }
```
Comment (herebebeasties) [inline System.IO.Abstractions.TestingHelpers/IMockFileDataAccessor.cs]: This feels slightly icky to me, but it was the least invasive change to make, as the various constructors for all the mock classes can then remain the same.

File: System.IO.Abstractions.TestingHelpers/MockDirectory.cs [hunk 2]
```diff
@@ -18,7 +18,7 @@ public class MockDirectory : DirectoryBase
 
         private string currentDirectory;
 
-        public MockDirectory(IMockFileDataAccessor mockFileDataAccessor, FileBase fileBase, string currentDirectory)
+        public MockDirectory(IMockFileDataAccessor mockFileDataAccessor, FileBase fileBase, string currentDirectory) : base(mockFileDataAccessor?.FileSystem)
```
Comment (herebebeasties) [inline System.IO.Abstractions.TestingHelpers/MockDirectory.cs]: The reason the above feels a bit icky is that you end up with these `null` checks which don't make all that much sense. The alternative here is obviously to pass in both the `IMockFileDataAccessor` and the `IFileSystem` to all these. Would people prefer we do that instead?

File: System.IO.Abstractions/DirectoryInfoBase.cs [hunk 3]
```diff
@@ -102,7 +106,7 @@ public abstract class DirectoryInfoBase : FileSystemInfoBase
         {
             if (directoryInfo == null)
                 return null;
-            return new DirectoryInfoWrapper(directoryInfo);
+            return new DirectoryInfoWrapper(new FileSystem(), directoryInfo);
```
Comment (herebebeasties) [inline System.IO.Abstractions/DirectoryInfoBase.cs]: I can't see much alternative to this given how the implicit cast works.
We could have a readonly static `FileSystem.Instance` field which we encourage people to use, I guess.

(To be honest I'm not a huge fan of implicit casts like this - they make dependency injection harder and tend to hide bugs, as witnessed by the one I fixed as part of this PR in `MockFileInfo`. Part of the public API, though. :-/)

File: System.IO.Abstractions/FileSystem.cs [hunk 4]
```diff
@@ -30,26 +36,25 @@ public IFileStreamFactory FileStream
...
         }
 
         DirectoryInfoFactory directoryInfoFactory;
         public IDirectoryInfoFactory DirectoryInfo
         {
-            get { return directoryInfoFactory ?? (directoryInfoFactory = new DirectoryInfoFactory()); }
+            get { return directoryInfoFactory ?? (directoryInfoFactory = new DirectoryInfoFactory(this)); }
         }
 
-        private readonly Lazy<DriveInfoFactory> driveInfoFactory = new Lazy<DriveInfoFactory>(() => new DriveInfoFactory());
```
Comment (herebebeasties) [inline System.IO.Abstractions/FileSystem.cs]: I'm not sure why this is the only one done as a `Lazy<T>`, but I've left it as-is.
I can fix this up to make is all consistent in another PR if you like?
(Or do it as part of this one if you don't mind the slight diff noise.)

File: System.IO.Abstractions.TestingHelpers/IMockFileDataAccessor.cs (hunk 1 above)
Comment (fgreinacher) [reply System.IO.Abstractions.TestingHelpers/IMockFileDataAccessor.cs]: I like it this way!

File: System.IO.Abstractions/FileSystem.cs (hunk 4 above)
Comment (fgreinacher) [reply System.IO.Abstractions/FileSystem.cs]: Would be great to make this consistent (I slightly prefer `Lazy<T>`)

File: System.IO.Abstractions/FileBase.cs [hunk 5]
```diff
@@ -8,6 +8,16 @@ namespace System.IO.Abstractions
     [Serializable]
     public abstract class FileBase
     {
+        protected FileBase(IFileSystem fileSystem)
+        {
+            this.FileSystem = fileSystem;
```
Comment (fgreinacher) [inline System.IO.Abstractions/FileBase.cs]: Nit: Redundant `this.`

File: System.IO.Abstractions/DirectoryBase.cs [hunk 6]
```diff
@@ -7,6 +7,16 @@ namespace System.IO.Abstractions
     [Serializable]
     public abstract class DirectoryBase
     {
+        protected DirectoryBase(IFileSystem fileSystem)
+        {
+            this.FileSystem = fileSystem;
```
Comment (fgreinacher) [inline System.IO.Abstractions/DirectoryBase.cs]: Nit: Redundant `this.`

File: System.IO.Abstractions/DriveInfoBase.cs [hunk 7]
```diff
@@ -3,6 +3,13 @@
     [Serializable]
     public abstract class DriveInfoBase
     {
+        protected DriveInfoBase(IFileSystem fileSystem)
+        {
+            FileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+        }
+
+        public IFileSystem FileSystem { get; }
```
Comment (fgreinacher) [inline System.IO.Abstractions/DriveInfoBase.cs]: Comment is missing

File: System.IO.Abstractions/FileSystem.cs [hunk 8]
```diff
@@ -3,22 +3,28 @@
     [Serializable]
     public class FileSystem : IFileSystem
     {
-        DirectoryBase directory;
+        public FileSystem()
+        {
+            driveInfoFactory = new Lazy<DriveInfoFactory>(() => new DriveInfoFactory(this));;
```
Comment (fgreinacher) [inline System.IO.Abstractions/FileSystem.cs]: Nit: Double `;` at the end

File: System.IO.Abstractions/FileSystemInfoBase.cs [hunk 9]
```diff
@@ -4,6 +4,13 @@
     [Serializable]
     public abstract class FileSystemInfoBase
     {
+        protected FileSystemInfoBase(IFileSystem fileSystem)
+        {
+            FileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+        }
+
+        public IFileSystem FileSystem { get; }
```
Comment (fgreinacher) [inline System.IO.Abstractions/FileSystemInfoBase.cs]: Missing comment

File: System.IO.Abstractions/FileSystem.cs (hunk 4 above)
Comment (herebebeasties) [reply System.IO.Abstractions/FileSystem.cs]: I've binned all the lazy entirely and just construct these instead - given all the constructors do basic field init (added in this PR) and nothing else, I really can't understand why these were made lazy in the first place!

**Comment by herebebeasties** [comment]
I've addressed all your comments, Florian. Thanks. Merged up to master, tests are all green.

**Comment by fgreinacher** [comment]
Great, thanks again!

**Pull request merged**